docs/advanced/pycpp/numpy.rst

```diff
@@ -205,6 +205,35 @@ prevent many types of unsupported structures, it is still the user's
 responsibility to use only "plain" structures that can be safely manipulated as
 raw memory without violating invariants.
 
+Scalar types
+============
+
+In some cases we may want to accept or return NumPy scalar values such as
+``np.float32`` or ``np.uint16``. It is especially important in case of
+C-side single-precision floats which by default will be bound to Python's
+double-precision builtin floats, causing mismatch in float precision.
+
+Luckily, there's a helper type for this occasion - ``py::numpy_scalar``:
+
+.. code-block:: cpp
+
+    m.def("square_float32", [](py::numpy_scalar<float> value) {
+        float v = value;
+        return py::make_scalar(v * v);
+    });
+
+This type is trivially convertible to and from the type it wraps; currently
```

Comment:
Why not the complex types as well?
https://docs.scipy.org/doc/numpy/user/basics.types.html

Reply:
`np.complex64` we probably can (`std::complex<float>`), but `np.complex128` is now platform-dependent (matches precision of built-in complex in Python).

Reply:
I would add `np.csingle` and `np.cdouble`, no?

Reply:
See the detailed comment below re: `csingle`, `cdouble`, `float` and `double`.